### DISCUSSION TITLE
real(1/100) and real(1*0.01) not the same?

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#21178] [Q&A] [answered]
Could some one explain this behavior?
```
print(real(1*0.01))
print(real(1/100))
0.01
0
```

**Comment by s-hadinger** [comment, marked as answer]
`1/100` is an integer operation. For float you need at least one of both to be a real like `1.0/100` or `1/100.0` or `1.0/100.0`

**Reply by sfromis** [reply]
... or `real(1) / 100` - etc.

When messing with floating-point numbers, get used to that 0.01 * 100 is not exactly 1, in the same way as 0.3333333 * 3 is not exactly 1.

**Reply by ChrisVonFrosta** [reply]
thanks to both.

i need one float on starting point, check.

Understood that rounding errors will happen. In my usecase this doesn't matter. i use this to convert the range 0-100 (%) frome a slider to 0.0-1.0. for the light controlling  method.

**Comment by Wintespe** [comment]
0.01 * 100 = 1.0
0.3333333 * 3 = 0.999999
May be different due to simplified implementation in Tasmota

**Reply by sfromis** [reply]
No. To check if it is **exactly** equal: `print(0.01*100, 0.01*100 == 1, 0.01*100 - 1)`
Result: `1 false 1.19209e-07`

Binary numbers which would have infinite digits after the decimal point when represented in base 10 will be rounded when converting to print format, which cannot be an exact representation of the bits of the number value.

My example of `0.3333333 * 3` was only illustrative about the reason why it is **not** exactly 1.0. The key point is "infinite decimals", whatever you see when printed is not the "whole truth". It was not meant to suggest a value where you'd see a visible or arithmetic difference. In some cases, internal rounding during calculation may even "happen" to gloss over the lost bits.

**Comment by Wintespe** [comment]
In c++ at  MS VisualStudio
...
	float f = 1.0;
	int i = 1;
	bool b = (i==f);
	f = i - f;
...
After execution b is true and f = 0.00000000.  Why?

**Reply by ChrisVonFrosta** [reply]
i have no glue but maybe the compiler optimization helps here?!

i would try something like this:
```
float f = 0.5
int i = 1;
f = (i - f)-f;
```
is f still 0.000000?

**Reply by sfromis** [reply]
Numbers having the non-integer part composed of negative powers of two can be represented 100% exactly in binary floating point (within the number of available bits, of course). Negative powers of two are 0.5, 0.25, 0.125, 0.0625, etc. Hence, `f` will be exact all the way.

When it comes to the example of Wintespe, it is even more straightforward, as as there are no decimals involved. Integer numbers can always be represented 100% exactly, again within the number of available bits. Hence the values are really the exact number.

The pertinent example from this thread is 0.01, which cannot be written as a sum of a finite number of negative powers of two, hence the inexactness I demonstrated.

**Reply by sfromis** [reply]
Here are a few numbers, and the internal [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) representation in hex:
```
1	3f800000
0.5	3f000000
0.25	3e800000
0.75	3f400000
0.1	3dcccccd
0.01	3c23d70a
```
When there's zeros to the right, the value is exact. Otherwise it is just the first of infinite bits trying to get as close as possible to the base-10 value. Similar to when 0.3333333 is not the same as 1/3.
